scylla/src/authentication/mod.rs

```diff
@@ -0,0 +1,96 @@
+    /// The information contained in the token parameter is authentication protocol specific.
+    /// It may be NULL or empty.
+    async fn evaluate_challenge(
+        &mut self,
+        token: Option<&[u8]>,
+    ) -> Result<Option<Vec<u8>>, AuthError>;
+
+    /// To handle the success phase of exchange.
+    /// The token parameters contain information that may be used to finalize the request.
+    async fn success(&mut self, token: Option<&[u8]>) -> Result<(), AuthError>;
+}
+
+/// Trait used to represent a factory of [`AuthenticatorSession`] instances.
+/// A new [`AuthenticatorSession`] instance will be created for each session.
+///
+/// The custom authenticator can be set using SessionBuilder::authenticator_provider method.  
+///
+/// Default: [`PlainTextAuthenticator`] is the default authenticator which requires username and
+/// password. It can be set by using SessionBuilder::user(\"user\", \"pass\") method.
+#[async_trait]
+pub trait AuthenticatorProvider: Sync + Send {
+    /// A pair of initial response and boxed [`AuthenticatorSession`]
+    /// should be returned if authentication is required by the server.
+    async fn start_authentication_session(
+        &self,
+        name: &str,
+    ) -> Result<(Option<Vec<u8>>, Box<dyn AuthenticatorSession>), AuthError>;
+}
+
+struct PlainTextAuthenticatorSession;
+
+#[async_trait]
+impl AuthenticatorSession for PlainTextAuthenticatorSession {
+    async fn evaluate_challenge(
+        &mut self,
+        _token: Option<&[u8]>,
+    ) -> Result<Option<Vec<u8>>, AuthError> {
+        Err("Challenges are not expected during PlainTextAuthentication".to_string())
+    }
+
+    async fn success(&mut self, _token: Option<&[u8]>) -> Result<(), AuthError> {
+        Ok(())
+    }
+}
+
+/// Default authenticator provider that requires username and password if authentication is required.
+pub struct PlainTextAuthenticator {
+    username: String,
+    password: String,
+}
+
+impl PlainTextAuthenticator {
+    /// Creates new [`PlainTextAuthenticator`] instance with provided username and password.
+    pub fn new(username: String, password: String) -> Self {
+        PlainTextAuthenticator { username, password }
+    }
+}
+
+#[async_trait]
+impl AuthenticatorProvider for PlainTextAuthenticator {
+    async fn start_authentication_session(
+        &self,
+        _name: &str,
+    ) -> Result<(Option<Vec<u8>>, Box<dyn AuthenticatorSession>), AuthError> {
+        let mut response = BytesMut::new();
+        let username_as_bytes = self.username.as_bytes();
+        let password_as_bytes = self.password.as_bytes();
+        let buf_size = 2 + username_as_bytes.len() + password_as_bytes.len();
+
+        // The body of response is a single [bytes] value (i32 length and then contents)
+        match buf_size.try_into() {
+            Ok(size) => response.put_i32(size),
+            Err(err) => return Err(err.to_string()),
+        }
```

Comment:
Hint: the length of the message is a part of the `[bytes]`. The AuthenticatorProvider should not be concerned with it.

Reply:
This code shouldn't be here. The `start_authentication_session` function should just return the [bytes] value (called `token` in the protocol specification). For password-based authentication, the token should not contain its length. The layer higher above the code is responsible for serializing the AuthResponse message, which requires serializing the token length and then its contents.

Even though the message is incorrect, Scylla accepts it. This is Scylla's fault as it is too permissive. It parses the token from authentication response from the back, first looking for password and then for the username and ignores the rest - so, if there is garbage at the beginning of the token then it will be ignored. Try running the authentication tests with Cassandra and you will see that it rejects it.

Reply:
Link to Scylla's token parsing code: https://github.com/scylladb/scylladb/blob/ff5527deb139275b9742e14bc62984caadd754f1/auth/sasl_challenge.cc#L30-L61